File: Day-10/labs/python-production-lab/capstone_producer.py
"""Capstone Python producer — same role as java CapstoneProducer."""

from __future__ import annotations

import json
import random
import sys

from kafka import KafkaProducer

from config import DEFAULT_BOOTSTRAP


def main() -> None:
    bootstrap = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_BOOTSTRAP
    topic = sys.argv[2] if len(sys.argv) > 2 else "capstone-orders"
    count = int(sys.argv[3]) if len(sys.argv) > 3 else 100

    producer = KafkaProducer(
        bootstrap_servers=bootstrap,
        acks="all",
        compression_type="lz4",
        linger_ms=5,
        batch_size=65536,
        value_serializer=lambda v: json.dumps(v).encode("utf-8"),
        key_serializer=lambda k: k.encode("utf-8"),
    )

    for i in range(count):
        order = {
            "orderId": f"ord-{i}",
            "customerId": f"cust-{random.randint(0, 19)}",
            "region": f"region-{random.randint(0, 3)}",
            "amount": 500 + random.randint(0, 9500),
            "valid": True,
        }
        producer.send(topic, key=order["orderId"], value=order)

    producer.flush()
    producer.close()
    print(f"Produced {count} events to {topic}")


if __name__ == "__main__":
    main()

File: Day-10/labs/capstone/java-capstone/src/main/java/com/training/kafka/capstone/CapstoneProducer.java
package com.training.kafka.capstone;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.Properties;
import java.util.Random;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

/**
 * Capstone starter — tuned producer emitting order events to capstone-orders.
 *
 * <p>Run: mvn -q exec:java -Dexec.mainClass=com.training.kafka.capstone.CapstoneProducer
 */
public final class CapstoneProducer {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Random RND = new Random();

    public static void main(String[] args) throws Exception {
        String bootstrap = args.length > 0 ? args[0] : "localhost:9092";
        String topic = args.length > 1 ? args[1] : "capstone-orders";
        int count = args.length > 2 ? Integer.parseInt(args[2]) : 100;

        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrap);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.ACKS_CONFIG, "all");
        props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "true");
        props.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, "lz4");
        props.put(ProducerConfig.LINGER_MS_CONFIG, "5");
        props.put(ProducerConfig.BATCH_SIZE_CONFIG, 65536);

        try (KafkaProducer<String, String> producer = new KafkaProducer<>(props)) {
            for (int i = 0; i < count; i++) {
                ObjectNode order = MAPPER.createObjectNode();
                order.put("orderId", "ord-" + i);
                order.put("customerId", "cust-" + RND.nextInt(20));
                order.put("region", "region-" + RND.nextInt(4));
                order.put("amount", 500 + RND.nextInt(9500));
                order.put("valid", true);
                String json = MAPPER.writeValueAsString(order);
                producer.send(new ProducerRecord<>(topic, order.get("orderId").asText(), json));
            }
            producer.flush();
        }
        System.out.println("Produced " + count + " events to " + topic);
    }
}
